File: tests/runone.m
function runone(filename,andquit)

if nargin<2
	andquit=1
end

addpath('./minFunc_2012');
addpath('./minFunc_2012/autoDif');
addpath('./minFunc_2012/minFunc');

fid = fopen(filename);
line = fgetl(fid);
funname = textscan(line,'%s %s');
funname = funname{1}{1};
line = fgetl(fid);
x0 = str2num(line);
x0 = x0';
options = [];
args = {};
procargs = false;
line = fgetl(fid);
while ischar(line)
	if isempty(line)
		procargs = true;
	elseif procargs
		args{end+1} = line;
	else
		vals = textscan(line,'%s %s');
		fname = vals{1}{1};
		fvalstr = vals{2}{1};
		[fval,isnum] = str2num(fvalstr);
		if isnum
			options = setfield(options,fname,fval);
		else
			options = setfield(options,fname,fvalstr);
		end
	end
	line = fgetl(fid);
end


fprintf('===START HERE\n')
if length(procargs)==0
	evalstr = ['minFunc(@' funname ',x0,options);'];
else
	evalstr = ['minFunc(@' funname ',x0,options' strjoin(args,',') ');'];
end
if ~isfield(options,'noutputs') || options.noutputs == 3
	eval(['[x,f,exitflag] = ' evalstr]);
	nout = 3;
elseif options.noutputs == 1
	eval(['x = ' evalstr]);
	nout = 1;
elseif options.noutputs == 2
	eval(['[x,f] = ' evalstr]);
	nout = 2;
else
	eval(['[x,f,exitflag,output] = ' evalstr]);
	nout = 4;
end

fprintf('x [ ');
fprintf('%.10g ',x(1:end-1));
fprintf('%.10g ]\n',x(end));
if nout>1
	fprintf('f %.10g\n',f);
end
if nout>2
	fprintf('exitflag %d\n',exitflag);
end
if nout>3
	fprintf('output.iterations %d\n',output.iterations);
	fprintf('output.funcCount %d\n',output.funcCount);
	fprintf('output.algorithm %d\n',output.algorithm);
	fprintf('output.firstorderopt %.10g\n',output.firstorderopt);
	fprintf('output.message %s\n',output.message);
	fprintf('output.trace.fval ');
	fprintf('%.10g ',output.trace.fval);
	fprintf('\n');
	fprintf('output.trace.funcCount ');
	fprintf('%d ',output.trace.funcCount);
	fprintf('\n');
	fprintf('output.trace.optCond ');
	fprintf('%.10g ',output.trace.optCond);
	fprintf('\n');
end
fprintf('===END HERE\n');
if andquit==1
	exit(0);
end

File: tests/runone.py
import sys
sys.path.insert(0,'../src')
from pyminfunc import minFunc
import numpy as np
from testfns import *

def runone(filename):
    with open(filename) as fid:
        line = fid.readline().rstrip()
        funname = line.split()[1]
        line = fid.readline().rstrip()
        x0 = np.fromstring(line,sep=' ')
        options = {}
        args = []
        procargs = False
        line = fid.readline().rstrip()
        while line:
            if line == '':
                procargs = true
            elif procargs:
                args.append(line)
            else:
                vals = line.split()
                fname = vals[0]
                fvalstr = vals[1]
                fval = vals[1]
                for type in (int,float):
                    try:
                        fval = type(fvalstr)
                    except ValueError:
                        continue
                options[fname] = fval
            line = fid.readline().rstrip()


    print('===START HERE')
    if len(args)==0:
        exestr = 'minFunc('+funname+',x0,options)'
    else:
        exestr = 'minFunc('+funname+',x0,options'+','.join(args)+')'

    if 'noutputs' not in options or options['noutputs'] == 3:
        x,f,exitflag = eval('minFunc('+funname+',x0,options)')
        output = None
    elif options['noutputs'] == 1:
        x = eval('minFunc('+funname+',x0,options)')
        f = None
        exitflag = None
        output = None
    elif options['noutputs'] == 2:
        x,f = eval('minFunc('+funname+',x0,options)')
        exitflag = None
        output = None
    elif options['noutputs'] == 4:
        x,f,exitflag,output = eval('minFunc('+funname+',x0,options)')

    print('x [ '+' '.join(['{:.10g}'.format(v) for v in x])+' ]')
    if f is not None: print(f'f {f:.10g}')
    if exitflag is not None: print('exitflag',exitflag)
    if output is not None:
        print('output.iterations',output.iterations)
        print('output.funcCount',output.funcCount)
        print(f'output.algorithm {output.algorithm:d}')
        print(f'output.firstorderopt {output.firstorderopt:.10g}')
        print('output.message',output.message)
        print('output.trace.fval',' '.join(['{:.10g}'.format(v) for v in output.trace.fval])+' ')
        print('output.trace.funcCount',' '.join(['{:d}'.format(v) for v in output.trace.funcCount])+' ')
        print('output.trace.optCond',' '.join(['{:.10g}'.format(v) for v in output.trace.optCond])+' ')
    print('===END HERE\n')

runone(sys.argv[1])
